### PR TITLE
NIAD-378: ACG-acceptance--gp-to-ha-birth-mandatory

### PR DESCRIPTION
**Pull request by bartek-sarul**

### CODE REVIEW

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/service/FhirToEdifactService.java [hunk 1]
```diff
@@ -79,7 +78,7 @@ private String getHaCipher(Patient patient) throws FhirValidationException {
 
     private String getRecipientTradingPartnerCode(Patient patient) throws FhirValidationException {
         String haCipher = getHaCipher(patient);
-        if(haCipher.length() == 2) {
+        if (haCipher.length() == 2) {
```
Comment (grzegorz-robak) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/service/FhirToEdifactService.java:81]: Is this code correct? If haCipher has 4 chars then we'll add 5th one

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonDateOfBirthMapper.java [hunk 2]
```diff
@@ -17,6 +21,12 @@ public PersonDateOfBirth map(Parameters parameters) {
 
     private Instant getPersonDob(Parameters parameters) {
         Patient patient = ParametersExtension.extractPatient(parameters);
-        return patient.getBirthDate().toInstant();
+        var birthDateElement = patient.getBirthDateElement();
```
Comment (bartek-sarul) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonDateOfBirthMapper.java:24]: Fhir parser sets local timezone when parsing json
I don't know how to change that so I did this trick

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonDateOfBirthMapper.java [hunk 3]
```diff
@@ -17,6 +21,12 @@ public PersonDateOfBirth map(Parameters parameters) {
...
+        return LocalDateTime
+            .of(
+                birthDateElement.getYear(), birthDateElement.getMonth() + 1, birthDateElement.getDay(),
```
Comment (bartek-sarul) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonDateOfBirthMapper.java]: I love the +1 on 0-indexed month

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonNameMapper.java [hunk 4]
```diff
@@ -15,6 +17,7 @@ public PersonName map(Parameters parameters) {
 
         return PersonName.builder()
             .nhsNumber(getNhsNumber(patient))
+            .patientIdentificationType(PersonName.PatientIdentificationType.OPI)
```
Comment (bartek-sarul) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonNameMapper.java]: We will have to handle OPI or API when doing the amendments

**Review by jaroslaw-wojciechowski — DISMISSED** [review]

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonDateOfBirthMapper.java (hunk 3 above)
Comment (mattd-kainos) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonDateOfBirthMapper.java]: Does it make you HAPI?

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonDateOfBirthMapper.java [hunk 5]
```diff
@@ -17,6 +21,12 @@ public PersonDateOfBirth map(Parameters parameters) {
...
+                birthDateElement.getHour(), birthDateElement.getMinute(), birthDateElement.getSecond())
```
Comment (mattd-kainos) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonDateOfBirthMapper.java]: Could you use LocalDate and not include the hour, minute, and second?

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/translator/AcceptanceBirthTranslator.java [hunk 6]
```diff
@@ -1,18 +1,83 @@
...
     public List<Segment> translate(Parameters parameters) throws FhirValidationException {
-        return null;
+
+        return Stream.of(
+            //BGM
+            emptyMapper(new BeginningOfMessage()),
+            //NAD+FHS
+            partyQualifierMapper,
+            //DTM+137
+            emptyMapper(new DateTimePeriod(timestampService.getCurrentTimestamp(), DateTimePeriod.TypeAndFormat.TRANSLATION_TIMESTAMP)),
```
Comment (mattd-kainos) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/translator/AcceptanceBirthTranslator.java]: Don't provide a timestamp here. FhirToEdifactService will do it for you. Ensures we have the same timestamp in the header and the body.

File: src/test/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonDateOfBirthMapperTest.java [hunk 7]
```diff
@@ -1,35 +1,31 @@
...
-        6,
-        23,
-        55,
-        0,
-        0,
-        ZoneId.of("Europe/London")).toInstant();
+        1991, 11, 6, 23, 55, 0, 0, ZoneId.of("Europe/London"))
+        .toInstant();
+    private static final Instant FIXED_TIME_LOCAL = ZonedDateTime.of(
+        1991, 11, 6, 23, 55, 0, 0, ZoneId.systemDefault())
```
Comment (mattd-kainos) [inline src/test/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonDateOfBirthMapperTest.java]: Why did you change this to systemDefault? Will we get a different results depending on the system's time zone?

**Review by mattd-kainos — CHANGES_REQUESTED** [review]
just a few minor things

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonDateOfBirthMapper.java (hunk 5 above)
Comment (bartek-sarul) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonDateOfBirthMapper.java]: I could but the "birthDateElement" has an option to have time as well. Should we trim that ?

File: src/test/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonDateOfBirthMapperTest.java (hunk 7 above)
Comment (bartek-sarul) [reply src/test/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonDateOfBirthMapperTest.java]: It won't. This is needed to pass tests in non-UK zone as fhir parser takes local zone

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonDateOfBirthMapper.java (hunk 5 above)
Comment (bartek-sarul) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonDateOfBirthMapper.java]: Ok, it's only date. Fixing

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/PersonName.java [hunk 8]
```diff
@@ -126,4 +125,15 @@ public void preValidate() throws EdifactValidationException {
     @Override
     protected void validateStateful() throws EdifactValidationException {
     }
+
+    public enum PatientIdentificationType {
+        OPI("Official Patient Identifier"),
```
Comment (mattd-kainos) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/PersonName.java]: Should be the other way around OFFICIAL_PATIENT_IDENTIFIER("OPI")

1) "description" string services no purpose here
2) if someone were to refactor the enum names it would break messaging

**Review by mattd-kainos — CHANGES_REQUESTED** [review]
just the enum name/description thing

**Review by grzegorz-robak — APPROVED** [review]

**Review by jaroslaw-wojciechowski — APPROVED** [review]

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/service/FhirToEdifactService.java (hunk 1 above)
Comment (mattd-kainos) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/service/FhirToEdifactService.java:81]: I think its ok. In that case GP System sent the wrong data, and we not validating content at this point.

**Review by mattd-kainos — APPROVED** [review]
